### ISSUE TITLE
nim accepts invalid code when template instantiation fails

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#12749]
nim accepts invalid code when template instantiation fails.

### Example
```nim
  type Bar[T: static string] = object
    x: int

  proc baz(b: string): string =
    echo ("baz", b)
    doAssert b.len > 0 # BUG: this is ignored
    return b
  proc main() =
    let a1 = Bar[baz("hi")](x: 12) # ok
    echo a1
    echo ("a1T", a1.T)
    let a2 = Bar[baz("")](x: 12) # BUG: should give CT error
    echo a2
    echo ("a2T", a2.T)
  main()
```

### Current Output
```
(x: 12)
("a1T", "hi")
(x: 12)
("a2T", "")
```

### Expected Output
* should give CT error: b.len > 0 assert failure because of 2nd call `let a2 = Bar[baz("")](x: 12)`
* also, `echo ("baz", b)` should not be gagged, but that's a separate issue

### Additional Information
* bug since 0.20 and in latest devel 80b508d33727a5b2fd8a3abd49df004e7d5c5138
* in 0.19, gives another buggy behavior:
there is not CT error (there should be, that's a bug); and there's instead a runtime error; furthermore, `echo ("baz", b)` is gagged for the 1st call but not for the 2nd call:
```
(x: 12)
(Field0: "a1T", Field1: "hi")
(x: 12)
(Field0: "baz", Field1: "")
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0805.nim(47) t0805
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0805.nim(44) main
/Users/timothee/.choosenim/toolchains/nim-0.19.0/lib/t0805.nim(38) baz
/Users/timothee/.choosenim/toolchains/nim-0.19.0/lib/system.nim(3790) failedAssertImpl
/Users/timothee/.choosenim/toolchains/nim-0.19.0/lib/system.nim(3783) raiseAssert
/Users/timothee/.choosenim/toolchains/nim-0.19.0/lib/system.nim(2830) sysFatal
Error: unhandled exception: /Users/timothee/git_clone/nim/timn/tests/nim/all/t0805.nim(38, 14) `
0 < len(b)`  [AssertionError]
```

## note
I'm using this pattern
```
  type Bar[T: static string] = object
    x: int
```
in a more complex use case as the basis of a very nice feature I discovered, which allows **schema free typesafe runtime configuration**; I can explain in another post / in forum.
In short it allows one to have a json(or any other config format) config file, and use it as follows:

mycfg.json:
```json
{
  "foo":{
    "bar":{
      age: 12,
      address: "bob",
    }
  }
}
```

```nim
let mycfg = typesafeCfg("mycfg.json")
if mycfg.foo.bar:
  var x = mycfg.age
  doAssert x == 12
  mycfg.name = "new bob"
```
now the program, once compile, binds statically the types, but not the values, so we can re-run the binary without recompiling and it'll pick up new runtime values from an updated `mycfg.json`; if types mismatch, it'll give an error

**Comment by metagn** [comment]
Issue says "template instantiation" but means "generic type instantiation". No relation to Nim templates. Simpler case:

```nim
type Foo[T: static int] = object
  x: int

proc bar(): int =
  raise newException(CatchableError, "abc")
  123

let a = Foo[bar()](x: 456) # should fail
echo a # (x: 123)
echo ("aT", a.T) # ("aT", 456)
```

**Comment by metagn** [comment]
Related #22623
